cedar-policy-core/src/entities/json/value.rs

```diff
@@ -283,53 +283,113 @@ impl CedarValueJson {
-                                        BorrowedRestrictedExpr::new_unchecked(v),
-                                    )?,
-                                ))
-                            })
-                            .collect::<Result<_, JsonSerializationError>>()?,
-                    ))
-                }
+                check_for_reserved_keys(map.keys())?;
+                Ok(Self::Record(
+                    map.iter()
+                        .map(|(k, v)| {
+                            Ok((
+                                k.clone(),
+                                CedarValueJson::from_expr(
+                                    // assuming the invariant holds for `expr`, it must also hold here
+                                    BorrowedRestrictedExpr::new_unchecked(v),
+                                )?,
+                            ))
+                        })
+                        .collect::<Result<_, JsonSerializationError>>()?,
+                ))
             }
             kind => {
                 Err(JsonSerializationError::UnexpectedRestrictedExprKind { kind: kind.clone() })
             }
         }
     }
 
+    /// Convert a Cedar value into a `CedarValueJson`.
+    ///
+    /// Only throws errors in two cases:
+    /// 1. `value` is (or contains) a record with a reserved key such as
+    ///     "__entity"
+    /// 2. `value` is (or contains) an extension value, and the argument to the
+    ///     extension constructor that produced that extension value can't
+    ///     itself be converted to `CedarJsonValue`. (Either because that
+    ///     argument falls into one of these two cases itself, or because the
+    ///     argument is a nontrivial residual.)
+    pub fn from_value(value: Value) -> Result<Self, JsonSerializationError> {
+        match value {
+            Value::Lit(lit) => Ok(Self::from_lit(lit)),
+            Value::Set(set) => Ok(Self::Set(
+                set.iter()
+                    .cloned()
+                    .map(Self::from_value)
+                    .collect::<Result<_, _>>()?,
+            )),
+            Value::Record(map) => {
+                // if `map` contains a key which collides with one of our JSON
+                // escapes, then we have a problem because it would be interpreted
+                // as an escape when being read back in.
+                check_for_reserved_keys(map.keys())?;
+                Ok(Self::Record(
+                    map.iter()
+                        .map(|(k, v)| Ok((k.clone(), Self::from_value(v.clone())?)))
+                        .collect::<Result<JsonRecord, JsonSerializationError>>()?,
+                ))
+            }
+            Value::ExtensionValue(ev) => {
+                let ext_fn: &Name = &ev.constructor;
+                Ok(Self::ExtnEscape {
+                    __extn: FnAndArg {
+                        ext_fn: ext_fn.to_string().into(),
+                        arg: match ev.args.as_slice() {
```

Comment:
Does this really only work for extensions that only call one argument?

Reply:
Yes.  Our JSON formats only support extension functions that have exactly one argument.  This is currently fine because we do not have any extension functions with zero or 2+ arguments that are legal in restricted expressions.  Put another way, all of our extension _constructors_ have exactly one argument.